src/notation/view/internal/percussionnotepopupcontentmodel.cpp

```diff
@@ -0,0 +1,206 @@
+    : QObject{parent}
+{
+}
+
+void PercussionNotePopupContentModel::init()
+{
+    interaction()->shadowNoteChanged().onNotify(this, [this]() {
+        emit shouldShowButtonsChanged();
+        emit percussionNoteNameChanged();
+        emit keyboardShortcutChanged();
+    });
+}
+
+void PercussionNotePopupContentModel::prevDrumNote()
+{
+    const Drumset* ds = currentDrumset();
+    mu::engraving::ShadowNote* shadowNote = currentShadowNote();
+    IF_ASSERT_FAILED(ds && shadowNote) {
+        return;
+    }
+
+    const int currNote = currentDrumPitch();
+    const int currLine = ds->line(currNote);
+
+    int pitch = currNote - 1;
+    while (pitch != currNote) {
+        if (pitch < 0) {
+            // Wrap around
+            pitch = mu::engraving::DRUM_INSTRUMENTS - 1;
+        }
+        if (ds->isValid(pitch) && ds->line(pitch) == currLine) {
+            noteInput()->setDrumNote(pitch);
+            interaction()->showShadowNote(shadowNote->pos());
+            interaction()->selectionChanged().notify(); // TEMPORARY HACK - triggers AbstractNotationPaintView::scheduleRedraw
+            emit percussionNoteNameChanged();
+            emit keyboardShortcutChanged();
+            return;
+        }
+        --pitch;
+    }
+
+    UNREACHABLE;
+}
+
+void PercussionNotePopupContentModel::nextDrumNote()
+{
+    const Drumset* ds = currentDrumset();
+    mu::engraving::ShadowNote* shadowNote = currentShadowNote();
+    IF_ASSERT_FAILED(ds && shadowNote) {
+        return;
+    }
+
+    const int currNote = currentDrumPitch();
+    const int currLine = ds->line(currNote);
+
+    int pitch = currNote + 1;
+    while (pitch != currNote) {
+        if (pitch == mu::engraving::DRUM_INSTRUMENTS) {
+            // Wrap around
+            pitch = 0;
+        }
+        if (ds->isValid(pitch) && ds->line(pitch) == currLine) {
+            noteInput()->setDrumNote(pitch);
+            interaction()->showShadowNote(shadowNote->pos());
+            interaction()->selectionChanged().notify(); // TEMPORARY HACK - triggers AbstractNotationPaintView::scheduleRedraw
+            emit percussionNoteNameChanged();
+            emit keyboardShortcutChanged();
+            return;
+        }
+        ++pitch;
+    }
+
+    UNREACHABLE;
+}
+
+bool PercussionNotePopupContentModel::shouldShowButtons() const
+{
+    const Drumset* ds = currentDrumset();
+    if (!ds || currentDrumPitch() < 0) {
+        return false;
+    }
+
+    const int line = ds->line(currentDrumPitch());
+
+    int drumsOnSameLine = 0;
+    for (int pitch = 0; pitch < mu::engraving::DRUM_INSTRUMENTS; ++pitch) {
+        if (ds->isValid(pitch) && ds->line(pitch) == line) {
+            ++drumsOnSameLine;
+        }
+        if (drumsOnSameLine > 1) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+QString PercussionNotePopupContentModel::percussionNoteName() const
+{
+    const Drumset* ds = currentDrumset();
+    if (!ds || currentDrumPitch() < 0) {
```

Comment:
We should call currentDrumPitch() only once. Its implementation looks quite complex